src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -741,30 +740,65 @@ protected Task FireOnCompleted()
             lock (_onCompletedSync)
             {
                 onCompleted = _onCompleted;
-                _onCompleted = null;
             }
 
-            if (onCompleted == null)
+            if (onCompleted == null || onCompleted.Count == 0)
             {
                 return Task.CompletedTask;
             }
 
-            return FireOnCompletedAwaited(onCompleted);
+            return FireOnCompletedMayAwait(onCompleted);
         }
 
-        private async Task FireOnCompletedAwaited(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
+        private Task FireOnCompletedMayAwait(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
         {
-            foreach (var entry in onCompleted)
+
+            var count = onCompleted.Count;
+            for (var i = 0; i < count; i++)
```

Comment:
@Tratcher pointed out that registering a new callback inside a callback could be weird with this change. It looks to me that for each new registration inside a callback would cause one additional callback at the bottom of the stack not be invoked.

Reply:
Is this something we care about and should have a test for on all servers?

Reply:
We should at least know what would happen, and it would be good to be consistent across servers. Throwing or skipping the late registration would be best.

Reply:
I don't care about it that much (It's definitely a "you're holding it wrong" situation IMO), but this failure mode is particularly bad.

Reply:
What's the desired behavior? This is also already broken for OnStarting no? This code was copied from that pattern.

Reply:
Ideally, we'd throw which didn't happen before. Before we wouldn't fire the callbacks that were added too late which was at least better than not firing the previously-added callbacks.

Reply:
As long as it doesn't affect the performance.

Reply:
Another idea would be to execute callback directly inline if they are run during the event.

Reply:
Wouldn't it change the ordering?